12-offer-encoding.md

```diff
@@ -0,0 +1,1057 @@
+
+## Payment Proofs and Payer Proofs
+
+Note that the normal lightning "proof of payment" can only demonstrate that an
+invoice was paid (by showing the preimage of the `payment_hash`), not who paid
+it.  The merchant can claim an invoice was paid, and once revealed, anyone can
+claim they paid the invoice, too.[1]
+
+1. Sharing the minimum information required to prove sections of the
+   invoice in dispute (e.g. the description of the item, the payment
+   hash, and the merchant's signature).
+2. Contain a transferable proof that the user is the one who was
+   responsible for paying the invoice in the first place.
+
+Providing a key in *invoice_request* allows a user to prove that they were the one
+to request the invoice.  In addition, the Merkle construction of the BOLT 12
+invoice signature allows the user to selectively reveal fields of the invoice
+in case of dispute.
+
+# Encoding
+
+Each of the forms documented here are in
+[TLV](01-messaging.md#type-length-value-format) format.
+
+The supported ASCII encoding is the human-readable prefix, followed by a
+`1`, followed by a bech32-style data string of the TLVs in order,
+optionally interspersed with `+` (for indicating additional data is to
+come).
+
+## Requirements
+
+Readers of a bolt12 string:
+- if it encounters a `+` followed zero or more whitespace characters between 
+  two bech32 characters:
+  - MUST remove the `+` and whitespace.
+
+## Rationale
+
+The use of bech32 is arbitrary, but already exists in the bitcoin
+world.  We currently omit the six-character trailing checksum: QR
+codes have their own checksums, bech32 doesn't protect against many
+length differences, and bech32m is not yet widely supported.
+
+The use of `+` (which is ignored) allows use over limited
+text fields like Twitter:
+
+```
+lno1xxxxxxxx+
+
+yyyyyyyyyyyy+
+
+zzzzz
+```
+
+See [format-string-test.json](bolt12/format-string-test.json).
+
+## Signature Calculation
+
+All signatures are created as per
+[BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki),
+and tagged as recommended there.  Thus we define H(`tag`,`msg`) as
+SHA256(SHA256(`tag`) || SHA256(`tag`) || `msg`), and SIG(`tag`,`msg`,`key`)
+as the signature of H(`tag`,`msg`) using `key`.
+
+Each form is signed using one or more TLV signature elements; TLV
+types 240 through 1000 are considered signature elements.  For these
+the tag is `lightning` | `messagename` | `fieldname`, and `msg` is the
+Merkle-root; `lightning` is the literal 9-byte ASCII string,
+`messagename` is the name of the TLV stream being signed (i.e. `offer`, `invoice_request` or `invoice`) and the `fieldname` is the TLV field containing the
+signature (e.g. `signature` or `payer_signature`).
+
+The formulation of the Merkle tree is similar to that proposed in
+[BIP-taproot], with the each TLV leaf paired with a nonce leaf to avoid
+revealing adjacent nodes in proofs (assuming there is a non-revealed TLV
+which has enough entropy).
```

Comment:
Most TLVs don't seem to have a lot of entropy. It seems to me that this complex construction will give people misplaced confidence that their message is private even though in many cases it will be easily guessable. Why not use a random nonce?

Reply:
Well, it turns out payer_info adds entropy in practice (if it doesn't add enough, you have other problems).  We could add another field for a random nonce, but you could just pick any odd field to do so.